tgui/packages/tgui/interfaces/PersonalCrafting.tsx

```diff
@@ -0,0 +1,318 @@
+import { filter, sortBy } from 'common/collections';
+import { flow } from 'common/fp';
+import { BooleanLike, classes } from 'common/react';
+import { useBackend } from '../backend';
+import { Button, Dimmer, Icon, Section, Stack, Table } from '../components';
+import { Window } from '../layouts';
+
+type RawRecipe = {
+  name: string;
+  ref: string;
+  req_text: string;
+  catalyst_text: string;
+  tool_text: string;
+};
+
+type RawData = {
+  // Dynamic data
+  busy: BooleanLike | null;
+  category: string;
+  subcategory: string;
+  display_craftable_only: BooleanLike;
+  display_compact: BooleanLike;
+  craftability: Record<string, BooleanLike>;
+  // Static data
+  crafting_recipes: Record<
+    string,
+    // That's the weirdest piece of type I had to write in my life.
+    // Yes, that's how subcategories are sent here.
+    RawRecipe[] | (Record<string, RawRecipe[]> & { has_subcats: 1 })
+  >;
+};
+
+type Recipe = RawRecipe & {
+  craftable: boolean;
+  dm_category: string;
+  dm_subcategory?: string;
+};
+
+type Category = {
+  name: string;
+  dm_category: string;
+  dm_subcategory?: string;
+};
+
+type Group = {
+  name: string;
+  categories: Category[];
+};
+
+type Data = RawData & {
+  dm_category: string;
+  dm_subcategory: string;
+  recipes: Recipe[];
+  groups: Group[];
+};
+
+const GENERAL_GROUP = 'General';
+
+const remapData = (rawData: RawData): Data => {
+  const craftability_by_ref = rawData.craftability;
+  const dm_categories = rawData.crafting_recipes || {};
+
+  const recipes: Recipe[] = [];
+  const generalGroup: Group = {
+    name: GENERAL_GROUP,
+    categories: [],
+  };
+  const groups: Group[] = [generalGroup];
+
+  for (const dm_category of Object.keys(dm_categories)) {
+    const dm_categories_item = dm_categories[dm_category];
```

Comment:
Why not Object.entries?

Reply:
I don't use it because it's heavily polyfilled and is slower.